navis/nbl/smat.py

```diff
@@ -0,0 +1,447 @@
+
+            Consider using ``numpy.logspace`` or ``numpy.geomspace``
+            to generate the inner boundaries of the bins.
+        dot_boundaries : array-like of floats
+            Monotonically increasing boundaries between
+            (possibly alpha-scaled) dot product bins
+            (i.e. N values makes N-1 bins).
+            If the first value > 0, -inf will be prepended,
+            otherwise the first value will be replaced with -inf.
+            If the last value < 1, inf will be appended,
+            otherwise the last value will be replaced with inf.
+
+            Consider using ``numpy.linspace`` between 0 and 1.
+        nonmatching : set of int, optional
+            Set of neurons, as indices into ``dotprops``,
+            which should not be considered matches.
+            If not given, all ``dotprops`` will be used
+            (on the assumption that matches are a small subset of possible pairs).
+        use_alpha : bool, optional
+            If true, multiply the dot product by the geometric mean
+            of the matched points' alpha values
+            (i.e. ``sqrt(alpha1 * alpha2)``).
+        seed : int, optional
+            Non-matching pairs are drawn at random using this seed,
+            by default {DEFAULT_SEED}
+        """
+        match_fn = dist_dot_alpha if use_alpha else dist_dot
+        super().__init__(
+            dotprops,
+            matching_sets,
+            [wrap_bounds(dist_boundaries, 0), wrap_bounds(dot_boundaries, 0, 1)],
+            match_fn,
+            nonmatching,
+            seed,
+        )
+
+    def build(self, threads=None) -> Lookup2d:
+        return Lookup2d(*self._build(threads))
+
+
+class LookupNd:
+    def __init__(self, boundaries: List[np.ndarray], cells: np.ndarray):
+        if [len(b) - 1 for b in boundaries] != list(cells.shape):
+            raise ValueError("boundaries and cells have inconsistent bin counts")
+        self.boundaries = boundaries
+        self.cells = cells
+
+    def __call__(self, *args):
+        if len(args) != len(self.boundaries):
+            raise TypeError(
+                f"Lookup takes {len(self.boundaries)} arguments but {len(args)} were given"
+            )
+
+        idxs = tuple(
+            np.digitize(r, b) - 1 for r, b in zip(args, self.boundaries)
+        )
+        out = self.cells[idxs]
+        return out
+
+
+def format_boundaries(arr):
+    return [f"[{lower},{upper})" for lower, upper in zip(arr[:-1], arr[1:])]
+
+
+def parse_boundary(item, strict=False):
+    explicit_interval = item[0] + item[-1]
+    if strict and item[0] + item[-1] != IMPLICIT_INTERVAL:
+        raise ValueError(
+            f"Enclosing characters {explicit_interval} "
+            f"do not match implicit interval specifiers {IMPLICIT_INTERVAL}"
+        )
+    return tuple(float(i) for i in item[1:-1].split(","))
+
+
+def parse_boundaries(items, strict=False):
+    last = None
```

Comment:
`last` is unnecessary, just use `upper`

Reply:
Still have to do a slightly long way round for type checking and 0-length reasons.